spec/services/feature_flag_spec.rb

```diff
@@ -3,40 +3,131 @@
+        FeatureFlag.activate(feature_name)
+        expect { FeatureFlag.deactivate(feature_name) }.to(
+          change { FeatureFlag.active?(feature_name) }.from(true).to(false),
+        )
+      end
+
+      it "records the change in the database" do
+        feature = Feature.create_or_find_by!(name: feature_name)
+        feature.update!(active: true)
+
+        expect { FeatureFlag.deactivate(feature_name) }.to(
+          change { feature.reload.active }.from(true).to(false),
+        )
+      end
+
+      it "does not deactivate object based flags" do
+        object = random_record
+        described_class.activate(feature_name, for: object)
+
+        expect { described_class.deactivate(feature_name) }
+          .not_to change { described_class.active?(feature_name, for: object) }
+      end
     end
 
-    it "records the change in the database" do
-      feature = Feature.create_or_find_by!(name: "add_participants")
-      feature.update!(active: true)
-      expect { FeatureFlag.deactivate(:add_participants) }.to(
-        change { feature.reload.active }.from(true).to(false),
-      )
+    context "with for: :all" do
+      it "deactivates a feature" do
+        FeatureFlag.activate(feature_name)
+        expect { FeatureFlag.deactivate(feature_name) }.to(
+          change { FeatureFlag.active?(feature_name) }.from(true).to(false),
+        )
+      end
+
+      it "records the change in the database" do
+        feature = Feature.create_or_find_by!(name: feature_name)
+        feature.update!(active: true)
+
+        expect { FeatureFlag.deactivate(feature_name) }.to(
+          change { feature.reload.active }.from(true).to(false),
+        )
+      end
+
+      it "does deactivate all object based flags" do
+        object = random_record
+        described_class.activate(feature_name, for: object)
+
+        expect { described_class.deactivate(feature_name, for: :all) }
+          .to change { described_class.active?(feature_name, for: object) }.to false
+      end
     end
+
+    context "with object passed via for: keyword" do
+      let(:object) { random_record }
+      let(:other_object) { random_record }
+
+      before do
+        described_class.activate(feature_name, for: object)
+        described_class.activate(feature_name, for: other_object)
+      end
+
+      it "does deactivate the feature for given object" do
+        expect { described_class.deactivate(feature_name, for: object) }
+          .to change { described_class.active?(feature_name, for: object) }.to false
+      end
+
+      it "does not deactivate the feature for other objects" do
+        expect { described_class.deactivate(feature_name, for: object) }
+          .not_to change { described_class.active?(feature_name, for: other_object) }
+      end
+    end
+  end
+
+  def random_record
+    # TODO: WHY ARE SOME FACTORIES BROKEN?
+    create FactoryBot.factories.map(&:name).without(:participant_band, :pupil_premium).sample
```

Comment:
Ummm...
I want some context xD
What is broken exactly?

Reply:
I guess some factories are never meant to be used by themselves, just as auxiliary things for other factories to use. Mostly out of general convenience.

Is it likely that we will want feature flags against anything other than users and schools? Maybe lead providers? Maybe this should be 'opt in' for factories and not 'opt out'?

Reply:
So `participant_band` and `pupil_premium` do not have enough attributes defined to be valid on creation. They do however look like a valid domain objects and we probably should fix those factories.

TBH, Ideally I would not be using any exisitng record in this test and use `with_model` or similar gem to mock AR class. The point is that this is to work with any AR object and I wanted to make that obvious in the test.